File: cpp/samchon/library/URLVariables.cpp
#include <API.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;

/* ------------------------------------------------------------
	CONSTRUCTORS
------------------------------------------------------------ */
URLVariables::URLVariables()
	: super()
{
}
URLVariables::URLVariables(const WeakString &flashVars)
	: super()
{
	vector<WeakString> &items = flashVars.split("&");
	for (size_t i = 0; i < items.size(); i++)
	{
		WeakString &item = items[i];
		size_t index = item.find("=");

		if(index == string::npos)
			continue;

		string &key = item.substr(0, index).str();
		string &value = decode(item.substr(index + 1));
		
		set(key, value);
	}
}

/* ------------------------------------------------------------
	URL ENCODING & DECONDING
------------------------------------------------------------ */
auto URLVariables::encode(const WeakString &wstr) -> string
{
	string res;
	res.reserve(wstr.size() * 3);

	for (size_t i = 0; i < wstr.size(); i++) 
	{
		unsigned char ch = wstr[i];

		if 
			(
				('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') 
				|| ('0' <= ch && ch <= '9') 
				|| ch == '-' || ch == '_' || ch == '.' || ch == '~' 
				|| ch == '@' || ch == ':' || ch == '/' || ch == '\\'
			)
			res.push_back(ch);
		else if (ch == ' ')
			res.append("%20");
		else
			res.append
			({
				'%',
				toHex( ch >> 4 ),
				toHex( ch & 0x0F )
			});
	}	
	return res;
}
auto URLVariables::decode(const WeakString &wstr) -> string
{
	string res;
	res.reserve(wstr.size());

	for (size_t i = 0; i < wstr.size(); i++)
	{
		const char ch = wstr[i];

		/*if (ch == '+')
		{
			res.append({' '});
		}
		else */
		if (ch == '%' && wstr.size() > i + 2)
		{
			char ch1 = fromHex(wstr[i + 1]);
			char ch2 = fromHex(wstr[i + 2]);
			char decoded = (ch1 << 4) | ch2;

			res.append({ decoded });
			i += 2;
		}
		else
			res.append({ch});
	}
	return res;
}

auto URLVariables::toHex(unsigned char ch) -> char
{
	static const char lookup[] = "0123456789ABCDEF";

	return lookup[ch];
}
auto URLVariables::fromHex(unsigned char ch) -> char
{
	if (ch <= '9' && ch >= '0')
		ch -= '0';
	else if (ch <= 'f' && ch >= 'a')
		ch -= 'a' - 10;
	else if (ch <= 'F' && ch >= 'A')
		ch -= 'A' - 10;
	else
		ch = 0;

	return ch;
}

/* ------------------------------------------------------------
	TO_STRING
------------------------------------------------------------ */
auto URLVariables::toString() const -> string
{
	string str = "";
	for(const_iterator it = begin(); it != end(); it++)
	{
		if(it != begin())
			str.append("&");

		str.append(it->first);
		str.append("=");
		str.append(encode(it->second));
	}
	return move(str);
}

File: ts/src/samchon/library/URLVariables.ts
/// <reference path="../API.ts" />

namespace samchon.library
{
	/**
	 * <p> URLVariables class is for representing variables of HTTP. </p>
	 * 
	 * <p> URLVariables class allows you to transfer variables between an application and server.
	 * When transfering, URLVariables will be converted to a URI string. </p>
	 * 
	 * <ul>
	 *	<li> URI: Uniform Resource Identifier </li>
	 * </ul>
	 * 
	 * @reference http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/URLVariables.html
	 * @author Migrated by Jeongho Nam <http://samchon.org>
	 */
	export class URLVariables
		extends std.HashMap<string, string>
	{
		/**
		 * Default Constructor.
		 */
		public constructor();

		/**
		 * <p> Construct from a URL-encoded string. </p>
		 * 
		 * <p> The {@link decode decode()} method is automatically called to convert the string to properties of the {@link URLVariables} object. </p>
		 * 
		 * @param str A URL-encoded string containing name/value pairs.
		 */
		public constructor(str: string);

		public constructor(str: string = "")
		{
			super();

			if (str != "")
				this.decode(str);
		}

		/**
		 * Converts the variable string to properties of the specified URLVariables object.
		 * 
		 * @param str A URL-encoded query string containing name/value pairs.
		 */
		public decode(str: string): void
		{
			this.clear();

			if (str.trim() == "")
				return;
			if (str.indexOf("?") != -1)
				str = str.substr(str.indexOf("?") + 1);

			var var_pairs: string[] = str.split("&");
			for (let i: number = 0; i < var_pairs.length; i++)
			{
				let equal_index: number = var_pairs[i].indexOf("=");
				let key: string;
				let value: string;

				if (equal_index == -1)
				{
					key = var_pairs[i];
					value = "";
				}
				else
				{
					key = var_pairs[i].substr(0, equal_index);
					name = decodeURIComponent(var_pairs[i].substr(equal_index + 1));
				}

				this.insert([key, name]);
			}
		}

		/**
		 * Returns a string containing all enumerable variables, in the MIME content encoding application/x-www-form-urlencoded.
		 */
		public toString(): string
		{
			let str: string = "";

			for (let it = this.begin(); !it.equal_to(this.end()); it = it.next())
			{
				if (!it.equal_to(this.begin()))
					str += "&";

				str += it.first + "=" + encodeURIComponent(it.second);
			}
			return str;
		}
	}
}